src/SFA.DAS.Forecasting.Commitments.Functions/NServicebusTriggerFunctions/ApprenticeshipCompletedFunction.cs

```diff
@@ -9,23 +14,54 @@ namespace SFA.DAS.Forecasting.Commitments.Functions.NServicebusTriggerFunctions
     public class ApprenticeshipCompletedFunction
     {
         private readonly IForecastingDbContext _forecastingDbContext;
+        private readonly IMapper _mapper;
+        private readonly ICommitmentsApiClient _commitmentsApiClient;
+        private readonly ILogger _logger;
 
-        public ApprenticeshipCompletedFunction(IForecastingDbContext forecastingDbContext)
+        public ApprenticeshipCompletedFunction(IForecastingDbContext forecastingDbContext, 
+            ICommitmentsApiClient commitmentsApiClient,
+            IMapper mapper,
+            ILogger<ApprenticeshipCompletedFunction> logger)
         {
             _forecastingDbContext = forecastingDbContext;
+            _commitmentsApiClient = commitmentsApiClient;
+            _mapper = mapper;
+            _logger = logger;
         }
 
         [FunctionName("ApprenticeshipCompleted")]
         public async Task Run(
             [NServiceBusTrigger(Endpoint = "SFA.DAS.Fcast.ApprenticeshipCompletedEvent")] ApprenticeshipCompletedEvent message)
         {
-            var selectedApprenticeship = _forecastingDbContext.Commitment.FirstOrDefault(x => x.ApprenticeshipId == message.ApprenticeshipId);
-            if (selectedApprenticeship != null)
+            try
             {
-                selectedApprenticeship.ActualEndDate = message.CompletionDate;
-                selectedApprenticeship.Status = Status.Completed;
+                _logger.LogInformation($"Apprenticeship Completed function executing at: [{DateTime.UtcNow}] UTC, event with ApprenticeshipId: [{message.ApprenticeshipId}].");               
+
+                var selectedApprenticeship = _forecastingDbContext.Commitment.FirstOrDefault(x => x.ApprenticeshipId == message.ApprenticeshipId);
+                if (selectedApprenticeship != null)
+                {
+                    selectedApprenticeship.ActualEndDate = message.CompletionDate;
+                    selectedApprenticeship.Status = Status.Completed;
+                }
+                else
+                {
+                    var apprenticeshipResponse = await _commitmentsApiClient.GetApprenticeship(message.ApprenticeshipId);              
+                    var result = _mapper.Map<Commitments>(apprenticeshipResponse);
+                    result.Status = Status.Completed;
+                    result.ActualEndDate = message.CompletionDate;
+                    _forecastingDbContext.Commitment.Add(result);
+                }
+
                 await _forecastingDbContext.SaveChangesAsync();
             }
+            catch(CommitmentsApiModelException commitmentException )
+            {
+                _logger.LogError(commitmentException, $"Apprenticeship Completed function Failure to retrieve  ApprenticeshipId: [{message.ApprenticeshipId}]");
```

Comment:
If there is an error then the function should re-throw so that is logs as a failure.

Reply:
done